src/main/scala/com/databricks/labs/mosaic/expressions/base/RasterBandExpression.scala

```diff
@@ -0,0 +1,59 @@
+package com.databricks.labs.mosaic.expressions.base
+
+import com.databricks.labs.mosaic.core.geometry.api.GeometryAPI
+import com.databricks.labs.mosaic.core.raster.{MosaicRaster, MosaicRasterBand}
+import com.databricks.labs.mosaic.core.raster.api.RasterAPI
+import com.databricks.labs.mosaic.functions.MosaicContext
+import org.apache.spark.sql.catalyst.expressions.{Expression, NullIntolerant, TernaryExpression}
+import org.apache.spark.sql.types.DataType
+import org.apache.spark.unsafe.types.UTF8String
+
+import scala.reflect.ClassTag
+
+abstract class RasterBandExpression[T <: Expression: ClassTag](
+    rasterExpr: Expression,
+    bandExpr: Expression,
+    pathExpr: Expression,
+    outputType: DataType
+) extends TernaryExpression
+      with NullIntolerant
+      with Serializable {
+
+    protected val geometryAPI: GeometryAPI = MosaicContext.geometryAPI
+
+    protected val rasterAPI: RasterAPI = MosaicContext.rasterAPI
+
+    override def first: Expression = rasterExpr
+
+    override def second: Expression = bandExpr
+
+    override def third: Expression = pathExpr
+
+    /** Output Data Type */
+    override def dataType: DataType = outputType
+
+    def bandTransform(raster: MosaicRaster, band: MosaicRasterBand): Any
+
+    override def nullSafeEval(inputRaster: Any, inputBand: Any, inputPath: Any): Any = {
+        val path = inputPath.asInstanceOf[UTF8String].toString
+        val bandIndex = inputBand.asInstanceOf[Int]
+
+        val baseRaster = rasterAPI.raster(inputRaster)
+        val raster = rasterAPI.raster(inputRaster, path)
```

Comment:
baseRaster and raster are of the same type. Why do we need two instances of the same raster?

Reply:
this is the funky virtualisation abstraction - I will add docs for this.
If the path is provided that means raster has subdatasets - ie. raster is a collection of rasters. 
So we do the outer raster then inner raster